### ISSUE TITLE
dest (accurate) function has bug

### ISSUE THREAD

**Issue by JamesGKent** [vtomanov/Gps64#1]
Firstly i want to say thank you for this library, it's been incredibly useful.

for a project i'm working on i needed to take a GPS fix and apply a bearing and distance offset to generate a new NMEA sentence at that offset, however i noticed a difference between your fast and accurate implementations for this (desto and dest) in that once the bearing goes beyond 180 degrees the output is offset in the wrong direction (instead of going west it is offset east) but with the correct north-south offset.

If i get to the bottom of it i'll add further details here.

**Comment by vtomanov** [comment]
OK I will take a look - but please use the desto version -a s the slow one
use the Float32 lib and there a bug in the libs trigonometric functions
which the guy owning the lib hasn't fixed yet...

please compare the algorithms with this site  :
https://www.movable-type.co.uk/scripts/latlong.html

and send a example values that are not correct and I will fix it ...

**Comment by JamesGKent** [comment]
so this is my sketch to generate NMEA sentences:
```
#include "Gps64.h"

GPS_DATA fix_lat;
GPS_DATA fix_lon;

GPS_DATA dst_lat;
GPS_DATA dst_lon;

void setup() {
	Serial.begin(57600);
	fix_lat.deg = 49;
	fix_lat.minSec = 685400000;
	fix_lon.deg = 6;
	fix_lon.minSec = 344266667;
}

void loop() {
	Serial.println("Start optimised");
	double distance = 100000.0;
        generate_sentence(fix_lat, fix_lon);
	for (double angle=0; angle<=360; angle += 10) {
		desto(fix_lat, fix_lon, angle, distance, dst_lat, dst_lon); // fast calc
		generate_sentence(dst_lat, dst_lon);
	}

	Serial.println("Start unoptimised");
	distance = 100000.0;
        generate_sentence(fix_lat, fix_lon);
	for (double angle=0; angle<=360; angle += 10) {
		if (angle == 190)
			distance = 110000.0;
		dest(fix_lat, fix_lon, angle, distance, dst_lat, dst_lon); // fast calc
		generate_sentence(dst_lat, dst_lon);
	}
	delay(30000);
}

int NMEAchecksum(const char *s) {
    int c = 0;
    while(*s)
        c ^= *s++;
    return c;
}

void generate_sentence(GPS_DATA latitude, GPS_DATA longitude) {
	int16_t lat_d = (latitude.deg >= 0) ? latitude.deg : latitude.deg*-1; // remove negative
	int16_t lat_m = ((latitude.minSec/100) * 6) / 1000000; // ((latitude.minSec * 6) / 100000000);
	int16_t lat_s = (((latitude.minSec/100) * 6) / 1000) - (lat_m * 1000); // (((latitude.minSec * 6) / 100000) - lat_m * 1000);
	char lat_h = (latitude.deg >= 0) ? 'N' : 'S';
	int16_t lon_d = (longitude.deg >= 0) ? longitude.deg : longitude.deg*-1; // remove negative
	uint8_t lon_m = ((longitude.minSec/100) * 6) / 1000000; // ((longitude.minSec * 6) / 100000000);
	uint16_t lon_s = (((longitude.minSec/100) * 6) / 1000) - (lon_m * 1000); // (((longitude.minSec * 6) / 100000) - lon_m * 1000);
	char lon_h = (longitude.deg >= 0) ? 'E' : 'W';
	
	static uint8_t offset;
	static char sentence_buf[64];
	offset = 0;
	memset(sentence_buf, 0, 64);
	
	offset = sprintf(sentence_buf, "$GPGGA,012345,");
	offset += sprintf(&sentence_buf[offset], "%02u%02u.%03u,%c,", lat_d, lat_m, lat_s, lat_h);
	offset += sprintf(&sentence_buf[offset], "%03u%02u.%03u,%c,", lon_d, lon_m, lon_s, lon_h);
	offset += sprintf(&sentence_buf[offset], "1,12,1.0,0.0,M,0.0,M,,");
	int chk = NMEAchecksum(&sentence_buf[1]); // offset to skip dollar sign
	offset += sprintf(&sentence_buf[offset], "*%02X", chk);
	Serial.println(sentence_buf);
}
```

which generates the following sets of outputs:
optimised:
```
$GPGGA,012345,4941.093,N,00620.655,E,1,12,1.0,0.0,M,0.0,M,,*71
$GPGGA,012345,4941.093,N,00620.670,E,1,12,1.0,0.0,M,0.0,M,,*76
$GPGGA,012345,4941.090,N,00620.684,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.086,N,00620.697,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.081,N,00620.709,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4941.074,N,00620.719,E,1,12,1.0,0.0,M,0.0,M,,*71
$GPGGA,012345,4941.066,N,00620.728,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.058,N,00620.734,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.049,N,00620.738,E,1,12,1.0,0.0,M,0.0,M,,*7C
$GPGGA,012345,4941.039,N,00620.739,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4941.030,N,00620.738,E,1,12,1.0,0.0,M,0.0,M,,*72
$GPGGA,012345,4941.021,N,00620.734,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.012,N,00620.728,E,1,12,1.0,0.0,M,0.0,M,,*73
$GPGGA,012345,4941.005,N,00620.719,E,1,12,1.0,0.0,M,0.0,M,,*77
$GPGGA,012345,4940.998,N,00620.709,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4940.993,N,00620.697,E,1,12,1.0,0.0,M,0.0,M,,*77
$GPGGA,012345,4940.989,N,00620.684,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4940.986,N,00620.670,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4940.985,N,00620.655,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4940.986,N,00620.641,E,1,12,1.0,0.0,M,0.0,M,,*78
$GPGGA,012345,4940.989,N,00620.627,E,1,12,1.0,0.0,M,0.0,M,,*77
$GPGGA,012345,4940.993,N,00620.614,E,1,12,1.0,0.0,M,0.0,M,,*7C
$GPGGA,012345,4940.998,N,00620.602,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.005,N,00620.592,E,1,12,1.0,0.0,M,0.0,M,,*76
$GPGGA,012345,4941.012,N,00620.583,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.021,N,00620.577,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.030,N,00620.573,E,1,12,1.0,0.0,M,0.0,M,,*7F
$GPGGA,012345,4941.039,N,00620.572,E,1,12,1.0,0.0,M,0.0,M,,*77
$GPGGA,012345,4941.049,N,00620.573,E,1,12,1.0,0.0,M,0.0,M,,*71
$GPGGA,012345,4941.058,N,00620.577,E,1,12,1.0,0.0,M,0.0,M,,*75
$GPGGA,012345,4941.066,N,00620.583,E,1,12,1.0,0.0,M,0.0,M,,*73
$GPGGA,012345,4941.074,N,00620.592,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.081,N,00620.602,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.086,N,00620.614,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.090,N,00620.627,E,1,12,1.0,0.0,M,0.0,M,,*77
$GPGGA,012345,4941.093,N,00620.641,E,1,12,1.0,0.0,M,0.0,M,,*74
$GPGGA,012345,4941.093,N,00620.655,E,1,12,1.0,0.0,M,0.0,M,,*71
```
accurate:
```
$GPGGA,012345,4941.177,N,00620.655,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4941.177,N,00620.670,E,1,12,1.0,0.0,M,0.0,M,,*7D
$GPGGA,012345,4941.174,N,00620.684,E,1,12,1.0,0.0,M,0.0,M,,*75
$GPGGA,012345,4941.170,N,00620.697,E,1,12,1.0,0.0,M,0.0,M,,*73
$GPGGA,012345,4941.165,N,00620.709,E,1,12,1.0,0.0,M,0.0,M,,*71
$GPGGA,012345,4941.158,N,00620.719,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.150,N,00620.728,E,1,12,1.0,0.0,M,0.0,M,,*74
$GPGGA,012345,4941.142,N,00620.734,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4941.133,N,00620.738,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.123,N,00620.739,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.114,N,00620.738,E,1,12,1.0,0.0,M,0.0,M,,*75
$GPGGA,012345,4941.105,N,00620.734,E,1,12,1.0,0.0,M,0.0,M,,*79
$GPGGA,012345,4941.097,N,00620.728,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.089,N,00620.719,E,1,12,1.0,0.0,M,0.0,M,,*73
$GPGGA,012345,4941.082,N,00620.709,E,1,12,1.0,0.0,M,0.0,M,,*79
$GPGGA,012345,4941.077,N,00620.697,E,1,12,1.0,0.0,M,0.0,M,,*75
$GPGGA,012345,4941.073,N,00620.684,E,1,12,1.0,0.0,M,0.0,M,,*73
$GPGGA,012345,4941.070,N,00620.670,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.070,N,00620.655,E,1,12,1.0,0.0,M,0.0,M,,*7C
$GPGGA,012345,4941.065,N,00620.671,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.068,N,00620.687,E,1,12,1.0,0.0,M,0.0,M,,*7A
$GPGGA,012345,4941.072,N,00620.701,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.078,N,00620.714,E,1,12,1.0,0.0,M,0.0,M,,*70
$GPGGA,012345,4941.085,N,00620.726,E,1,12,1.0,0.0,M,0.0,M,,*73
$GPGGA,012345,4941.094,N,00620.735,E,1,12,1.0,0.0,M,0.0,M,,*71
$GPGGA,012345,4941.103,N,00620.742,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.113,N,00620.746,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.123,N,00620.747,E,1,12,1.0,0.0,M,0.0,M,,*79
$GPGGA,012345,4941.134,N,00620.746,E,1,12,1.0,0.0,M,0.0,M,,*7E
$GPGGA,012345,4941.144,N,00620.742,E,1,12,1.0,0.0,M,0.0,M,,*7D
$GPGGA,012345,4941.153,N,00620.735,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.162,N,00620.726,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.169,N,00620.714,E,1,12,1.0,0.0,M,0.0,M,,*71
$GPGGA,012345,4941.175,N,00620.701,E,1,12,1.0,0.0,M,0.0,M,,*78
$GPGGA,012345,4941.179,N,00620.687,E,1,12,1.0,0.0,M,0.0,M,,*7B
$GPGGA,012345,4941.182,N,00620.671,E,1,12,1.0,0.0,M,0.0,M,,*76
$GPGGA,012345,4941.183,N,00620.655,E,1,12,1.0,0.0,M,0.0,M,,*71
```

i've been using the following tool to plot the outputs:
[https://nmeagen.org/](https://nmeagen.org/)

which gives me these outputs:
![optimised](https://user-images.githubusercontent.com/11315221/35848582-8fa4373e-0b16-11e8-8064-01a055c8ca0d.PNG)
![accurate](https://user-images.githubusercontent.com/11315221/35848583-8fd95068-0b16-11e8-9525-efb13b43bebd.PNG)


note that I added the offset in distance to the accurate version to make it easier to see the points doubling back.

I've tested with a few different locations now (around europe) and keep seeing the same result.
I am now using the optimised version so it's no longer critical to me that it gets updated, if as you say it's a bug in an upstream library then there probably isn't that much you can do about it.

**Comment by JamesGKent** [comment]
also not that the "accurate" version is centered in the right place, but the "optimised" version has all of the points south of whats supposed to be the centre.

**Comment by JamesGKent** [comment]
as a further note, because the offset appears to be the right amount but in the wrong direction, the following appears to give corrected results:
```
dest(fix_lat, fix_lon, angle, distance, dst_lat, dst_lon); // accurate calc
if (angle > 180) {
	int32_t diff = dst_lon.minSec - fix_lon.minSec;
	dst_lon.minSec -= (2*diff);
	if (dst_lon.minSec < 0) { // if overflow
		dst_lon.deg--;
		dst_lon.minSec += 1000000000;
	}
}
```

**Comment by vtomanov** [comment]
sorry :) can you summarize ?  I'm happy to change the code - just give a
exact suggestion and I will change it :)

**Comment by JamesGKent** [comment]
I'll try, but basically I've noticed two things, firstly that the "optimised" function can be much more wrong that the code comments suggest.
![optimised](https://user-images.githubusercontent.com/11315221/35913820-e584aa52-0bf8-11e8-80c4-a073bf12d42c.PNG)
in the above image the point at the very top is the coordinate fed in as the start point, so in this case should be the centre of the circle, the supporting code is used faithfully recreates the NMEA sentence i feed in so i know the problem isn't there.
I know it's not going to be as accurate, however your code comments suggest to within 10-20m, this should be a circle of radius 100m, however the distance from the "centre" point to the nearest other point is ~57m, and the furthest is ~250m, so a far greater inaccuracy.
I have no idea how to fix this, i'll be honest.

the second problem is with the "accurate" function:
![accurate](https://user-images.githubusercontent.com/11315221/35913932-54af2f60-0bf9-11e8-8fec-e6dd54a42e36.PNG)
when the bearing is less than 180 degrees the offset calculated is extremely accurate, however when the bearing goes above 180 degrees the longitude offset goes in the opposite direction to the way it should go, i.e. it offsets east when it should offset west.
I added the change in distance at 190 degrees to make it clear that the track goes back on itself for that picture.

for what i'm doing i'm using this function:
```
inline void dest(GPS_DATA & latStart,
                 GPS_DATA & lngStart,

                 double & bearing,
                 double & distance,

                 GPS_DATA & latEnd,
                 GPS_DATA & lngEnd);
```

however inside the library this just converts the arguments to f64 and calls this function:
```
inline void dest64(f64 & latStart,
                   f64 & lngStart,

                   f64 & bearing,
                   f64 & distance,

                   f64 & latEnd,
                   f64 & lngEnd)
```

so the issue must be somewhere around line 871 to 873:
https://github.com/vtomanov/Gps64/blob/master/Gps64.h#L871

however i'm at a loss for how to debug it.
I know that for my purposes wrapping the `dest` function provides a fix, but that only fixes the wrapper, not the underlying implementation.

to put the fix in the wrapper:
```
inline void dest(GPS_DATA & latStart,
                 GPS_DATA & lngStart,

                 double & bearing,
                 double & distance,

                 GPS_DATA & latEnd,
                 GPS_DATA & lngEnd)
{
	f64 _latStart(gtof64(latStart));
	f64 _lngStart(gtof64(lngStart));

	f64 _bearing(dtof64(bearing));
	f64 _distance(dtof64(distance));

	f64 _latEnd;
	f64 _lngEnd;

	dest64(_latStart,
		 _lngStart,

		 _bearing,
		 _distance,

		 _latEnd,
		 _lngEnd);

	latEnd = f64tog(_latEnd);
	lngEnd = f64tog(_lngEnd);
	if (bearing > 180) { // to fix issue where offsets always apply east
		int32_t diff = lngEnd.minSec - lngStart.minSec;
		lngEnd.minSec -= (2*diff);
		if (lngEnd.minSec < 0) { // if overflow
			lngEnd.deg--;
			lngEnd.minSec += 1000000000;
		}
	}
}
```

however this is not my recommended solution.

I'm still looking into this as i really want to give back to a library that has helped me massively, i will be putting together a pull request to add a feature i implemented because i'm lazy (populate GPS_DATA directly from RawDegrees)

**Comment by vtomanov** [comment]
have you compare the results using this site :  https://www.movable-type.co.
uk/scripts/latlong.html

**Comment by JamesGKent** [comment]
using that website i have created a comparison (csv) of calculated results using this library vs the output from the website you linked:
```
Angle,Lat,Lon,Web Lat,Web Lon
Centre,49°41'09"N,006°20'65"E,
0.00,49°41'63"N,006°20'65"E,49°41′41″N,006°21′05″E
10.00,49°41'62"N,006°20'79"E,49°41′41″N,006°21′14″E
20.00,49°41'60"N,006°20'94"E,49°41′39″N,006°21′22″E
30.00,49°41'56"N,006°21'07"E,49°41′37″N,006°21′30″E
40.00,49°41'50"N,006°21'19"E,49°41′34″N,006°21′37″E
50.00,49°41'43"N,006°21'29"E,49°41′30″N,006°21′43″E
60.00,49°41'36"N,006°21'37"E,49°41′25″N,006°21′48″E
70.00,49°41'27"N,006°21'43"E,49°41′20″N,006°21′52″E
80.00,49°41'18"N,006°21'47"E,49°41′15″N,006°21′54″E
90.00,49°41'09"N,006°21'48"E,49°41′09″N,006°21′55″E
100.00,49°40'99"N,006°21'47"E,49°41′03″N,006°21′54″E
110.00,49°40'90"N,006°21'43"E,49°40′58″N,006°21′52″E
120.00,49°40'82"N,006°21'37"E,49°40′53″N,006°21′48″E
130.00,49°40'74"N,006°21'29"E,49°40′48″N,006°21′43″E
140.00,49°40'67"N,006°21'19"E,49°40′44″N,006°21′37″E
150.00,49°40'62"N,006°21'07"E,49°40′41″N,006°21′30″E
160.00,49°40'58"N,006°20'94"E,49°40′39″N,006°21′22″E
170.00,49°40'56"N,006°20'79"E,49°40′37″N,006°21′14″E
180.00,49°40'55"N,006°20'65"E,49°40′37″N,006°21′05″E
190.00,49°40'56"N,006°20'79"E,49°40′37″N,006°20′56″E
200.00,49°40'58"N,006°20'94"E,49°40′39″N,006°20′48″E
210.00,49°40'62"N,006°21'07"E,49°40′41″N,006°20′40″E
220.00,49°40'67"N,006°21'19"E,49°40′44″N,006°20′33″E
230.00,49°40'74"N,006°21'29"E,49°40′48″N,006°20′27″E
240.00,49°40'82"N,006°21'37"E,49°40′53″N,006°20′22″E
250.00,49°40'90"N,006°21'43"E,49°40′58″N,006°20′18″E
260.00,49°40'99"N,006°21'47"E,49°41′03″N,006°20′16″E
270.00,49°41'09"N,006°21'48"E,49°41′09″N,006°20′15″E
280.00,49°41'18"N,006°21'47"E,49°41′15″N,006°20′16″E
290.00,49°41'27"N,006°21'43"E,49°41′20″N,006°20′18″E
300.00,49°41'36"N,006°21'37"E,49°41′25″N,006°20′22″E
310.00,49°41'43"N,006°21'29"E,49°41′30″N,006°20′27″E
320.00,49°41'50"N,006°21'19"E,49°41′34″N,006°20′33″E
330.00,49°41'56"N,006°21'07"E,49°41′37″N,006°20′40″E
340.00,49°41'60"N,006°20'94"E,49°41′39″N,006°20′48″E
350.00,49°41'62"N,006°20'79"E,49°41′41″N,006°20′56″E
360.00,49°41'63"N,006°20'65"E,49°41′41″N,006°21′05″E
```

you'll notice that the `Lon` column is effectively mirrored above/below 180 degrees, whereas the `Web Lon` column does not.

**Comment by vtomanov** [comment]
hmmm can be that one is expecting from -180 to 180 and the other 0-360 ?

**Comment by JamesGKent** [comment]
tried with bearing -180 to 180:
```
Angle,Lat,Lon,Web Lat,Web Lon
Centre,49°41'09"N,006°20'65"E,
-180.00,49°40'55"N,006°20'65"E,49°40′37″N,006°21′05″E
-170.00,49°40'56"N,006°20'79"E,49°40′37″N,006°20′56″E
-160.00,49°40'58"N,006°20'94"E,49°40′39″N,006°20′48″E
-150.00,49°40'62"N,006°21'07"E,49°40′41″N,006°20′40″E
-140.00,49°40'67"N,006°21'19"E,49°40′44″N,006°20′33″E
-130.00,49°40'74"N,006°21'29"E,49°40′48″N,006°20′27″E
-120.00,49°40'82"N,006°21'37"E,49°40′53″N,006°20′22″E
-110.00,49°40'90"N,006°21'43"E,49°40′58″N,006°20′18″E
-100.00,49°40'99"N,006°21'47"E,49°41′03″N,006°20′16″E
-90.00,49°41'09"N,006°21'48"E,49°41′09″N,006°20′15″E
-80.00,49°41'18"N,006°21'47"E,49°41′15″N,006°20′16″E
-70.00,49°41'27"N,006°21'43"E,49°41′20″N,006°20′18″E
-60.00,49°41'36"N,006°21'37"E,49°41′25″N,006°20′22″E
-50.00,49°41'43"N,006°21'29"E,49°41′30″N,006°20′27″E
-40.00,49°41'50"N,006°21'19"E,49°41′34″N,006°20′33″E
-30.00,49°41'56"N,006°21'07"E,49°41′37″N,006°20′40″E
-20.00,49°41'60"N,006°20'94"E,49°41′39″N,006°20′48″E
-10.00,49°41'62"N,006°20'79"E,49°41′41″N,006°20′56″E
0.00,49°41'63"N,006°20'65"E,49°41′41″N,006°21′05″E
10.00,49°41'62"N,006°20'79"E,49°41′41″N,006°21′14″E
20.00,49°41'60"N,006°20'94"E,49°41′39″N,006°21′22″E
30.00,49°41'56"N,006°21'07"E,49°41′37″N,006°21′30″E
40.00,49°41'50"N,006°21'19"E,49°41′34″N,006°21′37″E
50.00,49°41'43"N,006°21'29"E,49°41′30″N,006°21′43″E
60.00,49°41'36"N,006°21'37"E,49°41′25″N,006°21′48″E
70.00,49°41'27"N,006°21'43"E,49°41′20″N,006°21′52″E
80.00,49°41'18"N,006°21'47"E,49°41′15″N,006°21′54″E
90.00,49°41'09"N,006°21'48"E,49°41′09″N,006°21′55″E
100.00,49°40'99"N,006°21'47"E,49°41′03″N,006°21′54″E
110.00,49°40'90"N,006°21'43"E,49°40′58″N,006°21′52″E
120.00,49°40'82"N,006°21'37"E,49°40′53″N,006°21′48″E
130.00,49°40'74"N,006°21'29"E,49°40′48″N,006°21′43″E
140.00,49°40'67"N,006°21'19"E,49°40′44″N,006°21′37″E
150.00,49°40'62"N,006°21'07"E,49°40′41″N,006°21′30″E
160.00,49°40'58"N,006°20'94"E,49°40′39″N,006°21′22″E
170.00,49°40'56"N,006°20'79"E,49°40′37″N,006°21′14″E
180.00,49°40'55"N,006°20'65"E,49°40′37″N,006°21′05″E

results are the same, still doesn't offset west.

**Comment by vtomanov** [comment]
Hmm formulas are exactly as in the web site... and up to the 10th digit the trigonometric functions were ok - will take a look also ...

**Comment by JamesGKent** [comment]
it looks to me like the atan264 function is at fault, under certain circumstances it should return a negative value, eg for 240 degrees in the rough values of the left and right input to atan264 are:
-0.00009 and 0.41863  
using this website:
https://www.medcalc.org/manual/atan2_function.php
the calculated result should be -0.000214986978
however the value i'm getting is ~ 0.00021 (not negative)

this has been annoyingly difficult to debug as for some reason if i try to print the intermediary values (or pretty much any f64) it seems to currupt memory and my mega restarts...
not sure if it's linked to what version of the arduino IDE i'm using? 1.6.9? but i'm assuming thats a problem for the upstream lib (float64)

**Comment by JamesGKent** [comment]
also interestingly your atan64o function gives negative values under those circumstances, but the unoptimised or accurate version doesn't.

**Comment by vtomanov** [comment]
the ATAN2 is from NVIDIA :

http://developer.download.nvidia.com/cg/atan2.html

// Fast astan possible with error less then 1.8e-7
inline f64 atan264o(const f64 & y, const f64 & x)
{
f64 t3(abs64o(x));
f64 t1(abs64o(y));
f64 t0(max64o(t3, t1));
t1 = min64o(t3, t1);
t3 = f64(1L) / t0;
t3 = t1 * t3;
f64 t4(t3 * t3);
t0 = - (f64(13480470L) / f64(1000000000L));
t0 = t0 * t4 + (f64(57477314L) / f64(1000000000L)) ;
t0 = t0 * t4 - (f64(121239071L) / f64(1000000000L));
t0 = t0 * t4 + (f64(195635925L) / f64(1000000000L));
t0 = t0 * t4 - (f64(332994597L) / f64(1000000000L));
t0 = t0 * t4 + (f64(999995630L) / f64(1000000000L));
t3 = t0 * t3;
t3 = (abs64o(y) > abs64o(x)) ? f64(0x3FF921FB, 0x54442D18) - t3 : t3;
t3 = (x < f64(0L)) ? (f64(0x3FF921FB, 0x54442D18) * f64(2L)) - t3 : t3;
t3 = (y < f64(0L)) ? f64(0L) - t3 : t3;
return t3;
}

**Comment by JamesGKent** [comment]
I guess there's not much to do here, the issue is in the atan264 function from the Math64 library.
f64 atan264(f64 y, f64 x)
{
  // Special cases
  if(y.isZero() && x.isZero()) return 0; // undefined
  if(x.isZero() && y.isNegative()) return pio2*f64(-1);
  if(x.isZero() && !y.isNegative()) return pio2;	
  if(y.isNegative() && x.isNegative()) return atan64(y/x)-pio2*f64(2);
  if(x.isNegative()) return atan64(y/x)+pio2*f64(2); // y>=zero

  return atan64(y/x); // x>zero
}

I'll report it there.

**Comment by vtomanov** [comment]
do you sue the optimized version with the "o" - or you are talking only for
the slow version ?

**Comment by JamesGKent** [comment]
if I change the `dest64` function (accurate) to use `atan264o` I get the behavior I expect.
the fault lies with the `atan264` function from the Math64 library, no in this library, although until that library is fixed I'd suggest adding a warning that it's faulty?
I'm currently looking into how to fix the other libs function.
I'm going to close this issue as it's an upstream library problem, I really appreciate the help and your responsiveness.

James

**Comment by vtomanov** [comment]
in my scenario I use only the functions with "o" and the end - the fast one
as the normal one are really slow - some can take up to 12sec !

and also I needed to build the Tread64 lib to be able to do the calculation
in background and in the same time on foreground to read sensor, manage the
motors etc..

**Comment by JamesGKent** [comment]
thats understood, and I appreciate the massive speedup you've been able to achieve at the loss of some accuracy, however for my current project these aren't accurate enough and the speed isn't critical hence using the "accurate" or non o versions.
I've since noticed there is a similar bug in the upstream asin64 function where it returns the right magnitude but wrong sign value, so if south of the equator the bearing function returns a value thats north (sign flip) i'm putting a fix for that too into my fork of the Math64 lib, unless it gets merged (the guy seems to be inactive? it might be worth pointing at my fork for the Math64 library?

**Comment by vtomanov** [comment]
yep - I think I will add the fork to my project also !

how inaccurate you find the fast one in meters ?  in my case it actually
works quite well - my one is a boat electric motor with auto anchor and
return to home function

**Comment by JamesGKent** [comment]
i'll be honest that it only seems to be out in latitude, the longitude seems to be quite accurate, but in my tests it can be out by 250-300m in latitude.
its consistent in that if I plot a distance away from the centre at 10 degree increments, it looks like a circle, but its off centre. ie expecting this:
![acc](https://user-images.githubusercontent.com/11315221/35979054-367543d6-0cdf-11e8-934e-3175ab58693b.PNG)
but getting this:
![fast](https://user-images.githubusercontent.com/11315221/35979059-3ba26ffa-0cdf-11e8-9262-1cef63869268.PNG)

in that case everything calculated is shifted south by about 150m, when south of the equator things end up shifted north of where they should be, having had a play with my own version:
```
inline void dest64_test(f64 & latStart, f64 & lngStart, f64 & bearing, f64 & distance, f64 & latEnd, f64 & lngEnd) {
	f64 _latStart(dtor64(latStart));
	f64 _lngStart(dtor64(lngStart));
	
	f64 _bearing(dtor64(bearing));
	
	//lat2: =ASIN(SIN(lat1)*COS(d/R) + COS(lat1)*SIN(d/R)*COS(brng))
	//lon2: =lon1 + ATAN2(SIN(brng)*SIN(d/R)*COS(lat1),COS(d/R)-SIN(lat1)*SIN(lat2))
	f64 d_r((distance / f64(1000L)) / f64(6371000L));

	f64 _sin_latStart = sin64o(_latStart);
	f64 _cos_latStart = cos64o(_latStart);

	f64 _sin_d_r = sin64o(d_r);
	f64 _cos_d_r = cos64o(d_r);

	f64 _latEnd(asin64(_sin_latStart * _cos_d_r + _cos_latStart * _sin_d_r * cos64o(_bearing)));
	f64 _lngEnd(_lngStart + atan264o(sin64o(_bearing) * _sin_d_r * _cos_latStart, _cos_d_r - _sin_latStart * sin64o(_latEnd)));
	
	latEnd = rtod64(_latEnd);
	lngEnd = rtod64(_lngEnd);
}
```
I can get accurate results with a total run time to draw the circle of ~120s instead of ~192s

which is a shame, because if that asin64o function was more accurate i'd be more than happy to use it.
that said, i'm nowhere near clever enough to set about writing one...

**Comment by JamesGKent** [comment]
interestingly i implemented this idea here:
https://dsp.stackexchange.com/a/25785
in the form of:
```
inline f64 asin64_test(const f64 & rhs) {
	return atan64(f64(1)/(sqrt64(f64(1)-rhs*rhs)));
}
```
tried with both sqrt64 and sqrt64o, all of which give pretty much the same results as the asin64o implementation that from my research is also adapted from nvidia code? but also with no improvement to runtime.
my guess is that the inaccurate asin64o function is a bigger problem than an inaccurate atan264o as the atan264o is used to apply a small offset to the existing longitude, whereas the expression using asin64 replaces the latitude term, so the %error is relative to the size of the earth rather than the size of the offset. so for small offsets using the accurate asin64 func with the fast atan264o func gives a performance increase with no meaningful error.

hence my slightly tweaked version above uses the optimised sin, cos and atan2 functions and the accurate asin function. I also put some of the results into temp variables as they are used twice, to save calculating them twice.

**Comment by vtomanov** [comment]
If you can get the error fixed with no loss in performance will be great :)  in my case I constantly adjust the course of the boat and a slow calculation will not do...  500msec is the max for calculation in my case ..

**Comment by JamesGKent** [comment]
funny you should say that, while i haven't been able to implement a fast asin with a higher accuracy, i have implemented a fix:
```
inline void dest64(f64 & latStart, f64 & lngStart, f64 & bearing, f64 & distance, f64 & latEnd, f64 & lngEnd) {
	f64 _latStart(dtor64(latStart));
	f64 _lngStart(dtor64(lngStart));
	
	f64 _bearing(dtor64(bearing));
	
	//lat2: =ASIN(SIN(lat1)*COS(d/R) + COS(lat1)*SIN(d/R)*COS(brng))
	//lon2: =lon1 + ATAN2(SIN(brng)*SIN(d/R)*COS(lat1),COS(d/R)-SIN(lat1)*SIN(lat2))
	f64 d_r((distance / f64(1000L)) / f64(6371000L));

	f64 _sin_latStart = sin64o(_latStart);
	f64 _cos_latStart = cos64o(_latStart);

	f64 _sin_d_r = sin64o(d_r);
	f64 _cos_d_r = cos64o(d_r);

	static f64 offset_lat = _latStart;
	static f64 offset=0;
	if ((offset == 0) || (abs64o(offset_lat - _latStart) > 0.00174533)) { // if no offset calc'd or more than 0.1 degree away from last calc'd offset start
		offset = asin64o(_sin_latStart) - _latStart; // with no distance or bearing cos(d_r)=1, all the other terms cancel out (sin(0) = 0)
		offset_lat = _latStart;
	}

	f64 _latEnd(asin64o(_sin_latStart * _cos_d_r + _cos_latStart * _sin_d_r * cos64(_bearing)));
	f64 _lngEnd(_lngStart + atan264o(sin64(_bearing) * _sin_d_r * _cos_latStart, _cos_d_r - _sin_latStart * sin64(_latEnd)));
	
	latEnd = rtod64(_latEnd-offset);
	lngEnd = rtod64(_lngEnd);
}
```

this runs my benchmark at pretty much the same speed as your optimised version (it basically is) but calculates the latitude error first and then offsets the output. to help keep the speed up it only recalculates the offset if the input latitude changes by more than a certain amount, in this case by 0.1 degrees.

and changing the inline sin and cos lat and d_r calcs to be pre calculated saves about 18ms per func, so i believe its about 75ms saving?

let me know what you think.

**Comment by vtomanov** [comment]
I added you as collaborator to the GPS64 repository :) free free to
change/update the lib :)

**Comment by JamesGKent** [comment]
thanks, this is the first time I've ever been added as a collaborator, as a self taught programmer i'm honoured that you thought my contributions were good enough.
I'm not sure what range you have been using these functions over (my application is less than 50K) but in my tests the desto function is now pretty much as accurate as the dest function and actually slightly faster than before(~8%) hopefully you'll try it out and get the same improvements.

I've also added a pull request with my test sketch i used for running different scenarios, but i'll leave it up to you if you wish to merge that.

**Comment by vtomanov** [comment]
Great :)

I will rebuild my project with the latest commits from you also :)

get a look on the Tread64 - can be very useful also :)

**Comment by JamesGKent** [comment]
worth noting aswell that the "accurate" version of acos from the Math64 library is hopelessly broken once the fixes for atan2 and asin are applied. so any of your functions that depend on them will be similarly broken.
however your optimized versions are not affected.

I did have a brief look however most of the stuff I do is linear in->out type pipelines instead of multitasking, apart from some string handling from the serial ports, for which I have my Serializer library.
at some point I plan to add support for the NeoHWSerial library to it so I can implement interrupt handlers for it, but not a priority at the moment.
